### ISSUE TITLE
Transitive dependencies for native builds

### ISSUE THREAD

**Issue by donalmacc** [gradle/gradle#801]
_Edited by @lacasseio to convert into an epic._

### Expected Behavior
As a user, I can specify the native dependencies transitively for compiling, linking and installing.

For example:
Defining an executable that depends on library A that depends on library B should use library B as a dependency for the executable. 
* If A and B are shared libraries, then the resulting shared libraries should be copied as part of the install task
* If A and B are static libraries, when linking the executable, both A and B should be used to resolve symbols.

A better understanding of the scope of this work is explained in [this comment below](https://github.com/gradle/gradle/issues/801#issuecomment-258258496).

### Current Behavior
Each native component needs to specify its own dependencies and can't rely on transitivity.

This has been discussed in [an old forum post (august 2015)](https://discuss.gradle.org/t/native-cpp-automatically-handle-transitive-include-dependencies/11200) and has shown lots of interest since then.

A typical scenario is:
Defining an executable dependent on a library A, which itself depends on library B ignores library B when building the executable.

Two code examples extracted from real codebase have been attached to this issues:
* [Scenario 1: Dependencies are static libraries](https://github.com/gradle/gradle/files/569697/StaticLibDependency.zip)
Lib1 and Lib2 are static libraries. Exe depends on Lib1, Lib1 depends on Lib2. linking fails with unresolved symbols from Lib2 when linking Exe

* [Scenario 2: Dependencies are shared libraries](https://github.com/gradle/gradle/files/569696/SharedLibDependency.zip)
Lib1 and Lib2 are shared libraries  (Note: The sample use MSVC. The C++ is set up to export symbols from a DLL, so the code may not compile on gcc/clang). The sample correctly compiles, however after executing InstallExe1, the install directory only contains Lib1.dll instead of Lib1 and Lib2.

### Steps to Reproduce (for bugs)
See Current Behaviour, with full samples

### Your Environment
Windows 10
Gradle 3.1
MSVC 2013, MSVC 2015

**Comment by lacasseio** [comment]
Thanks @donalmacc for raising limitation. As you have noted, this is been discuss quite a lot in the pass. However, this is not a straight forward to implement as it seems. There is far more possible scenarios then the two exposed here. In the end, it all boils down to 3 type of dependencies:

1. **The `compiler` dependency.** This is the dependencies for the API (aka headers). It will be used by the selected toolchain compiler to transform the compilation unit (`*.c`, `*.cpp`, `*.cc`, etc.) into object files (`*.o`, `*.obj`, etc.).
2. **The `linker` dependency.** This is the dependencies for the implementation (aka static libraries as well as external object files). It will be used by the selected toolchain linker to transform the object files into an executable binary (ELF, PE, etc.). These executable binary are typically a shared libraries (`*.dll`, `*.so`, etc.) or an executables (`*.exe`, etc.).
3. **The `runtime` dependency.** This is the dependencies for the execution on the target hardware. It will be used either by the OS when resolving the binary import table or during code execution if it dynamically load binaries (with `LoadLibrary`, `dlopen`, etc.).

Those 3 dependency types can be totally independent, meaning we could compile against certain API (headers), link against someone else's implementation of those headers and, yet again, run with a different version of the linked library (like when a newer version of library is released simply addressing bug fix).

To add more to the mix, transitivity may not always be needed and depending on your compiler flag, this may create larger than expected binary if everything is transitive by default. For example, project A have a `linker` dependency to static library B which have a `compiler` dependency on the API (headers) of library C. We cannot assume project A will use any of the header from B nor C as it's dependent on the `#include` hierarchy of B used by A, if A use any of B's header in the first place. Remember that you could `extern` a symbol and not require any headers. Furthermore, we cannot assume project A will need library C during the linking phase as A may not even use any function from library B that require symbols from C to link successfully.

We can repeat similar scenario for the other type of dependencies. It's also important to note that projects in the wild may have stranger compilation requirement which can complexity this even more, like circular dependencies.

That been said it is not a challenge the Gradle engineer will shy away. We will address this limitation, it simply need time and additional thinking. This will likely be rolled out in phase. I would recommend anyone landing on this issue requiring such feature in Gradle to comment with your specific scenario (code example is important) and/or vote on previously commented scenario that apply to you using the +1 reaction so we get a better feel for what scenario happen the most often and prioritize according to those votes.

**Comment by donalmacc** [comment]
Hi Daniel, thanks for the quick response.

I didn't mean to imply that these were the only forms of dependencies, nor did I mean to imply that it was a simple task. I hadn't considered some of the complexities you mentioned; in particular extern symbols.

My samples weren't meant to be exhaustive, they were supposed to provide minimal examples of issues that we have hit in our codebase in the last week, which is what you've asked for other people to provide.

I'm not really in a position to comment on concerns about binary size/unnecessary dependencies, however this is a problem that other build tools (MSBuild, UnrealBuildTool, Ninja, scons)have all solved in one form or another. I don't doubt that there are projects out there with complex build requirements, but as it stands there's been 12 months since this was last discussed, and based on the reply in the other thread there hasn't been any movement towards implementing any sort of dependency management for the simplest cases, nor is there any documentation available about the intended solution for more complex dependencies. 

Also, I should probably mention that philmcardle (the person who bumped the forum post) and I both work on the same project.

**Comment by lacasseio** [comment]
No worries @donalmacc and thanks for the comment. I will see how we can get the work started on this.

**Comment by bigdaz** [comment]
@lacasseio and others, you might be interested in [this spike](https://github.com/gradle/gradle/tree/dd-better-native-dependency-resolution) that @big-guy and worked on to convert our native plugins to use the core dependency resolution engine.

For now, this spike doesn't add much functionality, but it gives us a clear path to improving native dependency resolution, including transitive project dependencies and external module dependencies.

**Comment by lacasseio** [comment]
Thanks a lot @big-guy and @bigdaz to have worked on this spike. We will definitely consider this work in this epic.

**Comment by lacasseio** [comment]
I will hijack this issue and convert it into an epic for better prioritization, rest assure @donalmacc no information will be lost.

**Comment by lacasseio** [comment]
Seeing the wonderful work been done by @philmcardle in the [referenced forum post](https://discuss.gradle.org/t/native-cpp-automatically-handle-transitive-include-dependencies/11200/10?u=daniel_l) made me remember another tricky part about transitive dependencies that I missed from my original comment: dependency order.

Native compilers have been around for a long time and they differ in features. For example, the link order is handled differently in Visual Studio vs GCC. Visual Studio tries to resolve the symbols as it goes through the libraries, keeping unreferenced symbol in memory where they can be referenced later during the linking phase. In contrast, GCC will discard any unreferenced symbol so they can't be referenced later during the linking phase.

I'm nowhere near an expert with all compilers and I know some flags can be used to work around some of those issues. It would be nice to have build experts comment on how this is addressed in complex build system to emulate similar behaviors in the transitive dependency implementation.

**Comment by martinda** [comment]
Could Ivy extra attributes be used explicitly or behind the scenes to correctly identify the artifacts of dependencies for a project of arbitrary type (not just native but other types of compilers as well)? Ivy configurations are great because they help me fetch the correct artifacts transitively, but I wish I could use Ivy extra attributes to refer to multiple dimensions of binary variants either in the `dependencies` closure or behind the scenes.

**Comment by big-guy** [comment]
Our new native plugins support transitive dependencies, binary dependencies and composite builds. https://github.com/gradle/gradle-native

**Comment by AnnDinKC** [comment]
Hello all –

We would like to not include transitive dependencies automatically at link time. Our reasoning is similar to what Fedora describes here around the dangers of transitive dependencies being pulled in: https://fedoraproject.org/wiki/UnderstandingDSOLinkChange 
e.g. If project A depends on library B & C and library B depends on C, project A can get away with only declaring a dependency on project B even though it needs project C. This locks project B into a passivity concern where it can no longer remove its dependency on project C without breaking project A. 

We would like to submit a PR to add a feature flag to disable transitive dependencies for CPP gradle at the link phase. 

Questions:
Would the community be open to feature flagging this ability? 

Would you point me to the diff where the changes around this epic were merged into gradle? I am having difficulty finding it by searching through closed pull issues and the 53k+ commits to the project.

Thank you for your time and consideration in this matter!

**Comment by big-guy** [comment]
Hi @AnnDinKC 

If I understand the problem correctly, I think we wouldn't suggest that you solve that problem by disabling all transitive dependencies. The ability to do this is already present. You can disable transitive dependencies for your dependencies via https://docs.gradle.org/nightly/javadoc/org/gradle/api/artifacts/Configuration.html#setTransitive-boolean-.

There are a couple use cases for transitive dependencies:
- `libfoo` has a dependency on `libbar` and to use `libfoo` you need to use types/functions defined in `libbar`. In this case, transitive dependencies do exactly what you want.
- `libfoo` has a dependency on `libbar`, but from your app's POV, you only deal with `libfoo`.  In this case, we don't want the headers or the libraries of `libfoo` when compiling and linking our app.  But you _may_ need the `libbar` libraries when running. In this case, transitive dependencies do what you want if Gradle knows that `libbar` is a private/implementation-only dependency of `libfoo`.

When declaring a dependency, Gradle lets you declare a dependency as `api` or `implementation`. An `api` dependency is carried downstream to other consumers and would usually represent the APIs that are required to use your library's API.  e.g., if you required Eigen matrix structures to be passed into your library's functions, you could say that eigen is an `api` dependency for your library.

An `implementation` dependency is a dependency that shouldn't be exposed to consumers. They may still need to be used for runtime (like shared libraries), but they're not required for compilation/linking. An example of this is if you use `log4cxx` internally, but you don't want to pollute your consumer's include path with `log4cxx`s headers.

I think that sort of separation would solve most of the problems of accidentally using dependencies of your dependencies because your include path would only include APIs from usable dependencies.

There are a couple of things this doesn't solve for compilation:
- If you have a common include path that contains multiple libraries, but you only declare a dependency on a subset, you may rely on something undeclared because the other undeclared headers are still reachable. This is a possibility due to the way #includes work. C++ modules might help here eventually.
- If you explicitly redefine (like `extern int foo();` in the example above), you may rely on something undeclared. I think this should be rare. 

In either case, linking problems should still occur with proper api/implementation separation.

While typing this up, I raised a new issue https://github.com/gradle/gradle-native/issues/759 -- we should be applying the same kind of restrictions on link-time visibility as we do for compile-time visibility when it comes to api/implementation dependencies.

HTH!  If you'd like more info, we have a separate Gradle Native project going over on https://github.com/gradle/gradle-native

**Comment by AnnDinKC** [comment]
@big-guy Thank you for the quick and thorough response! 

Before posting I had tried to use some of the  'Configuration' properties to configure transitive to be false, but concluded it was only for java builds and not native since I was not able to get it working. I am very new to gradle so it was likely misconfiguration.

Using `api` and `implementation` to differentiate dependencies may also solve the issues we are concerned about, and sounds like good practice to boot. I'll be working with both of these to try and get my teams builds where we want them.

Again, many thanks for your feedback. It has undoubtedly saved me many hours of going down the wrong rabbit hole.
